### ISSUE TITLE
Change Request: replace `natural-compare` with native one.

### ISSUE THREAD

**Issue by hyperz111** [eslint/eslint#20185]
### ESLint version

v9.36.0

### What problem do you want to solve?

Replace `natural-compare` with native one. See [this](https://github.com/eslint/eslint/blob/main/lib%2Frules%2Fsort-keys.js#L13-L13)

### What do you think is the correct solution?

Use this code instead:

```js
function naturalCompare(a, b) {
	return String.prototype.localeCompare.call(a, b, undefined, { numeric: true })
}
```

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by kecrily** [comment]
I've researched this before, and we can't simply replace `natural-compare` with `localeCompare`—their behaviors are inconsistent. We'd need to do some extra work, and I'm concerned the team might hesitate about maintaining it themselves.

Regarding the specific inconsistencies, I'll need to check them again first.

**Comment by kecrily** [comment]
```js
[
  '1',   '10', 'A1',
  'A10', 'Z1', 'a1',
  'a10', 'z1', 'À2',
  'É2',  'à1', 'é1'
]
```

The sorting order for natural-compare is numbers first, then letters, and finally letters with accent marks.

Attempting various combinations of localeCompare options fails to replicate the behavior of natural-compare.

```js
// String.prototype.localeCompare.call(a, b, 'en', { numeric: true, sensitivity: "variant", caseFirst: 'upper' })

[
  '1',   '10',  'A1',
  'a1',  'à1',  'À2',
  'A10', 'a10', 'é1',
  'É2',  'Z1',  'z1'
]
```

**Comment by aladdin-add** [comment]
I'm curious why it was used in the first commit - the differences seem so subtle that the vast majority of users wouldn't even use or notice them. Moreover, it's only used in this one rule. If that's the case, I'd prefer using the native because it makes ESLint smaller, faster to install, and more secure. would like to hear from others. / cc @eslint/eslint-team

**Comment by lumirlumir** [comment]
I agree with what @aladdin-add mentioned, and I'm personally +1 on replacing `natural-compare` with a native implementation.

Looking at the source code, it's less than 40 lines, and the test cases are pretty straightforward.

- Source code: https://github.com/litejs/natural-compare-lite/blob/main/index.js
- Tests: https://github.com/litejs/natural-compare-lite/blob/main/test/index.js

Switching to a native solution seems to have more benefits than drawbacks.

**Comment by DMartens** [comment]
I am also in favor of using a native implementation for the mentioned reasons but as this is likely a breaking change I would:
1. Check which of our test cases would change (current valid test cases may become invalid and vice versa)
2. Gather public eslintrcs which use the `natural` sorting and check what the differences would be
3. Add a temporary option which allows opting-in using `localeCompare` so users can check their code bases

I do not know if we still have time for getting this into v10.

**Comment by aladdin-add** [comment]
Although the features of v10 have been frozen, the impact should be minimal, and I hope to include it in v10. 

would like to have feedback from the tsc. / @eslint/eslint-tsc

**Comment by fasttime** [comment]
`natural-compare` is only used in the `sort-keys` rule, which is currently frozen.

> I'm curious why it was used in the first commit

According to the description in #6800, it's because [the corresponding JSCS rule](https://jscs-dev.github.io/rule/validateOrderInObjectKeys) was using `natural-compare`.

**Comment by fasttime** [comment]
I tried to replace the usages of `naturalCompare` as suggested in the issue:

https://github.com/eslint/eslint/blob/46eea6d1cbed41d020cb76841ebba30710b0afd0/lib/rules/sort-keys.js#L54-L59

When running the unit tests, that produced 18 failures, including for some really simple cases like `var obj = {c:2, b_:1, C:3}`.

I wouldn't be opposed to introducing a breaking change to remove a dependency, but in this case I'm afraid that the impact would be more than just minimal, so I'm not sure. Is there anything we can do to make the new behavior more similar to `natural-compare`?

**Comment by aladdin-add** [comment]
I replaced it with a simple implementation:


```js
		function naturalCompare(a, b) {
			if(/^\d+$/u.test(a) && /^\d+$/u.test(b)) {
				return Number(a) - Number(b);
			}
			return a <= b ? -1 : 1;
		}
```

it works for the case, and 211/215 tests passed✅.

**Comment by kecrily** [comment]
I'm not entirely sure if incompatibility is acceptable. For two reasons:

- naturalCompare feels more intuitive (though this may be subjective).
- The rule has already been frozen.

**Comment by JoshuaKGoldberg** [comment]
This is similar to #19696. Removing `natural-compare` provides near-zero benefit:

* Dependency tree size: `natural-compare` is 412B minified + gzipped (https://bundlephobia.com/package/natural-compare@1.4.0), which is a fraction of a fraction of a percent of ESLint's total size. This would not be noticeable to users.
* Runtime performance: roughly the same; not measurably different.

Meanwhile, per https://github.com/eslint/eslint/issues/20185#issuecomment-3415414290, we'd have to either:

* Keep to the proposed very minimal code -> result in a breaking change for users
* Add edge case handling logic to the added code -> we'd effectively be re-implementing `natural-compare`

On its own, I'm -1 on this change.

However, we should note that if there is a larger initiative to clean up ESLint for a bigger total goal (https://github.com/eslint/eslint/issues/19696#issuecomment-3426947393), that could be more convincing. I propose we wait for that proposal to go through before outright closing/rejecting this issue.

**Comment by nzakas** [comment]
Closing for now as we don't have a larger proposal to replace a bunch of proposals and this one isn't compelling on its own.
